### PR TITLE
Live data monitoring on reflectometry GUI

### LINKED ISSUE #18828: Formalise live data display
Add a button to the `ISIS Reflectometry` GUI to start live data monitoring
- It should run `ReflectometryReductionOneAuto` as a post processing algorithm
- It should use the instrument defined on the GUI
- It should get the angle and slits of the live run for use in RROA
- Properties on the Settings tab on the GUI should override the other inputs to RROA if they are specified. There are two groups on the GUI and it's unclear which should be used. It doesn't really matter so it will be fine to just assume that we always use Group1, as long as we make this clear.
- Ideally there should be an indicator on the GUI that live data is running
- Ideally live data monitoring should be restarted automatically if it drops out (which happens quite frequently when the data size changes)

We can get theta from ibex instruments using epics support in mantid. Reflectometry instruments are not on ibex yet but the controls team can install an additional process that will make the same variables available, e.g. this has been done for polref.
```
from epics import caget
title=caget('IN:POLREF:DAE:TITLE',as_string=True)
theta=caget('IN:POLREF:CS:SB:Theta',as_string=True)
print "Title:",title
print "Theta:",theta
```

### PR DESCRIPTION
**Pull request by gemmaguest**
This PR adds a button to start live data monitoring on the `ISIS Reflectometry` GUI. 

**Report to:** Max at ISIS

### Algorithms
Run this script. 
```
Load(Filename='INTER13460', OutputWorkspace='input')
ReflectometryReductionOneLiveData(Instrument='INTER', InputWorkspace='input', OutputWorkspace='live_output')
theta = GetLiveInstrumentValue(Instrument='INTER', PropertyType='Block',PropertyName='Theta')
ReflectometryReductionOneAuto(InputWorkspace='input', ThetaIn=theta)
```
- The theta value printed in the log should be the same as that shown on the [dashboard](http://dataweb.isis.rl.ac.uk/SeciWeb/DefaultImplement.aspx?Instrument=inter)
- Note that if the live theta value is zero the reduction should fail (if so, check it gives a sensible error).
- Ohterwise, the `live_output` workspace should contain a single spectrum similar to the `IvsQ_binned_13460` output from `ReflectometryReductionOneAuto` with the same `ThetaIn` (although note that the binning is different because it does not use the live slits)

live_vs_rroa

### GUI
These instructions assume that Theta is non-zero for INTER on the above dashboard. If it's not you could try one of the other instruments.

- Open the `ISIS Reflectometry` GUI
- Select `INTER` and click `Start monitor`
- Note that the `Start monitor` and `Stop monitor` buttons will be greyed out while live data is starting.
- If there is any error (e.g. theta is zero or not found) then it will be displayed at this point and live data will not run. `Start monitor` will be re-enabled.
- Otherwise, once live data is running, `Stop monitor` will be re-enabled.
- Open the Algorithm progress dialog (from the `Running...` button) and you should see that `StartLiveData` is running (if you open it quick enough) and then this should change to `MonitorLiveData`
- There should be two output workspaces, `TOF_live` and `IvsQ_binned_live`. The latter should have been reduced to a single spectrum.
- You should be able to stop the monitor from the GUI or the algorithms dialog. 

### Errors
- If the reduction fails due to invalid settings it should give a sensible error in the log but MonitorLiveData will continue trying to run it (so you will see the same error over and over again). A way to test this for OFFSPEC is to set the AnalysisMode to PointDetectorAnalysis and clear the ProcessingInstructions. There will be two output workspaces `TOF_live` and `IvsQ_binned_live`. They will both be the same because the reduction failed.
- We might encounter other errors due to the instrument setup. Sme errors (such as a `DAE` error or unfound EPICS values) can take a long time to produce if they have a long timeout. The monitor buttons on the GUI should remain greyed out until the error if this happens during startup. If it happens while the monitor is running, the monitor will stop. You can get a DAE error if you make the network connection very slow as described in PR #23443, and then wait till it times out.

Fixes #18828 

Does this update require release notes?
- [x] Yes
- [ ] No

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/PythonInterface/plugins/algorithms/GetLiveInstrumentValue.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+    def _name_prefix(self):
+        """Prefix to use in the EPICS string before the property name"""
+        if self._propertyType == 'Run':
+            return ':DAE:'
+        else:
+            return ':CS:SB:'
+
+    def _get_live_value(self):
+        from epics import caget
```
Comment (Mantid-Matthew) [inline Framework/PythonInterface/plugins/algorithms/GetLiveInstrumentValue.py]: Could we surround this import statement by a try/catch block to provide a more useful error message?

File: Framework/PythonInterface/plugins/algorithms/GetLiveInstrumentValue.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        """Prefix to use in the EPICS string before the property name"""
+        if self._propertyType == 'Run':
+            return ':DAE:'
+        else:
+            return ':CS:SB:'
+
+    def _get_live_value(self):
+        try:
+            from epics import caget
+        except:
```
Comment (Mantid-Matthew) [inline Framework/PythonInterface/plugins/algorithms/GetLiveInstrumentValue.py]: Could I be picky here and ask that we use a more specific exception type? Minor point but I think it is better to state the specific exception you are expecting where possible. For example in this case I think `except ImportError:` should work. 

Would it also be worth pointing them towards the relevant documentation here?

**Review by Matthew-Andrew — CHANGES_REQUESTED** [review]
Looks good to me, the testing worked as expected and code changes good and well tested. One minor picky point but should then be good to go.

**Review by Matthew-Andrew — APPROVED** [review]
All looks good to me

**Pull request merged**